app/src/main/java/com/kleshchin/danil/filemanager/MainActivity.java

```diff
@@ -0,0 +1,135 @@
+                    if (mimeTypeMap.hasExtension(
+                            mimeTypeMap.getFileExtensionFromUrl(uri.toString())))
+                        mime = mimeTypeMap.getMimeTypeFromExtension(
+                                mimeTypeMap.getFileExtensionFromUrl(uri.toString()));
+                    intent.setDataAndType(uri, mime);
+                    startActivity(intent);
+                }
+            }
+        });
+    }
+
+    private void initToolbar() {
+        toolbar = (Toolbar) findViewById(R.id.main_toolbar);
+        setSupportActionBar(toolbar);
+        getSupportActionBar().setDisplayShowTitleEnabled(false);
+        toolbar.setLogo(new ColorDrawable(getResources().getColor(android.R.color.transparent)));
+        if (!((TextView) findViewById(R.id.toolbar_title)).getText().equals("") &&
+                !((TextView) findViewById(R.id.toolbar_title)).getText().equals(mainPath)) {
+            final Drawable upArrow = getResources().getDrawable(R.drawable.ic_arrow_back_black_36dp);
+            upArrow.setColorFilter(getResources().getColor(R.color.colorWhite), PorterDuff.Mode.SRC_ATOP);
+            toolbar.setLogo(upArrow);
+            setLogoOnClickListener(toolbar, new View.OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    fillListView(new File(((TextView) findViewById(R.id.toolbar_title)).getText().
+                            toString().substring(0, ((TextView) findViewById(R.id.toolbar_title)).
+                            getText().toString().length() - 1 - currentDirectory.length() + 1)));
+                }
+            });
+        }
+    }
+
+    private void setLogoOnClickListener(Toolbar toolbar, View.OnClickListener listener) {
+        try {
+            Class<?> toolbarClass = Toolbar.class;
+            Field logoField = toolbarClass.getDeclaredField("mLogoView");
+            logoField.setAccessible(true);
+            ImageView logoView = (ImageView) logoField.get(toolbar);
+            if (logoView != null) {
+                logoView.setOnClickListener(listener);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void fillListView(File file) {
+        listItems.clear();
+        listFilesItems.clear();
+        listView.setAdapter(listAdapter);
+        File list[] = file.listFiles();
+        ((TextView) findViewById(R.id.toolbar_title)).setText(file.getPath());
+        initToolbar();
+        currentDirectory = file.getName();
+        if (list != null) {
+            for (File currentFile : list) {
+                if (currentFile.isDirectory()) {
+                    listItems.add(new ListItems(currentFile.getName(), R.drawable.folder));
+                } else {
+                    listFilesItems.add(new ListItems(currentFile.getName(), R.drawable.file));
+                }
+            }
+            listItems.addAll(listFilesItems);
```

Comment:
It's very suspicious that you need to have both listItems and listFilesItems collections. Perhaps you need to revise the logic to get rid of one of them